### PR TITLE
[Feature]: Benchmark of applying ViT to Mask-RCNN

### PR DESCRIPTION
**Pull request by YuanLiuuuuuu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/models/backbones/vit.py [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+            norm_cfg=norm_cfg,
+            final_norm=final_norm,
+            with_cls_token=with_cls_token,
+            output_cls_token=output_cls_token,
+            interpolate_mode=interpolate_mode,
+            patch_cfg=patch_cfg,
+            init_cfg=init_cfg)
+
+        # stochastic depth decay rule
+        dpr = np.linspace(0, drop_path_rate, self.arch_settings['num_layers'])
```
Comment (ZwwWayne) [inline mmdet/models/backbones/vit.py:342]: dpr -> drop_path_rates

File: configs/mask_rcnn/mask_rcnn_vit_base_fpn_coco.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            dict(use_window=True, window_size=14),
+            dict(use_window=False, window_size=14),
+            dict(use_window=True, window_size=14),
+            dict(use_window=True, window_size=14),
+            dict(use_window=False, window_size=14),
+        ],
+        init_cfg=dict(
+            type='Pretrained',
+            checkpoint=
+            'ckpt/deit-base_3rdparty_pt-16xb64_in1k_20211124-6f40c188.pth')),
```
Comment (jbwang1997) [inline configs/mask_rcnn/mask_rcnn_vit_base_fpn_coco.py:32]: Need to change to URL.

File: mmdet/models/backbones/vit.py [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+import math
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+from mmcls.models import VisionTransformer as _VisionTransformer
+from mmcls.models.backbones.vision_transformer import \
+    TransformerEncoderLayer as _TransformerEncoderLayer
+from mmcls.models.utils import MultiheadAttention as _MultiheadAttention
+from mmcls.models.utils import resize_pos_embed
```
Comment (jbwang1997) [inline mmdet/models/backbones/vit.py:11]: Users may not install the mmcls codebase.

File: mmdet/models/backbones/vit.py [hunk 4]
```diff
@@ -0,0 +1,458 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import math
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+from mmcls.models import VisionTransformer as _VisionTransformer
+from mmcls.models.backbones.vision_transformer import \
+    TransformerEncoderLayer as _TransformerEncoderLayer
+from mmcls.models.utils import MultiheadAttention as _MultiheadAttention
```
Comment (RangiLyu) [inline mmdet/models/backbones/vit.py:10]: Why not use the MultiheadAttention in mmcv.cnn.bricks.transformer?

**Comment by RangiLyu** [comment]
Move the config to configs/vit

File: configs/mask_rcnn/mask_rcnn_vit_base_fpn_coco.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
+_base_ = [
```
Comment (hhaAndroid) [inline configs/mask_rcnn/mask_rcnn_vit_base_fpn_coco.py:1]: You can describe the connection with `strong_baselines` in the configuration file. I see that most of the configuration is from `strong_baselines`.

Comment (hhaAndroid) [reply configs/mask_rcnn/mask_rcnn_vit_base_fpn_coco.py:1]: The configuration file name should also refer to the modification in `strong_baselines`

File: mmdet/models/backbones/vit.py (hunk 3 above)
Comment (hhaAndroid) [reply mmdet/models/backbones/vit.py:11]: try

File: mmdet/models/backbones/vit.py [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+from mmcv.cnn.bricks.transformer import FFN
+from mmcv.cnn.utils.weight_init import trunc_normal_
+from mmcv.runner.base_module import ModuleList
+from torch import nn
+
+from ..builder import BACKBONES
+
+
+class TransformerEncoderLayer(_TransformerEncoderLayer):
+    """Implements one encoder layer in Vision Transformer.
```
Comment (hhaAndroid) [inline mmdet/models/backbones/vit.py:23]: We should explain why we are rewriting here？

File: configs/mask_rcnn/mask_rcnn_vit_base_fpn_coco.py (hunk 5 above)
Comment (YuanLiuuuuuu) [reply configs/mask_rcnn/mask_rcnn_vit_base_fpn_coco.py:1]: What is `strong_baselines`

File: mmdet/models/backbones/vit.py [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+            B, C, *patch_resolution)
+        out1 = self.norm1(outs[1]).transpose(1, 2).reshape(
+            B, C, *patch_resolution)
+        out2 = self.norm2(outs[2]).transpose(1, 2).reshape(
+            B, C, *patch_resolution)
+        out3 = self.norm3(outs[3]).transpose(1, 2).reshape(
+            B, C, *patch_resolution)
+
+        # Convert to multi-scale
+        out0 = self.res_modify_block_0(out0).contiguous()
```
Comment (ZwwWayne) [inline mmdet/models/backbones/vit.py:524]: create a channel mapper module to map the feature maps to the expected shape and channels.

File: mmdet/models/backbones/vit.py [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+            num_extra_tokens=self.num_extra_tokens)
+        x = self.drop_after_pos(x)
+
+        if not self.with_cls_token:
+            # Remove class token for transformer encoder input
+            x = x[:, 1:]
+
+        outs = []
+        for i, layer in enumerate(self.layers):
+            rel_pos_bias = self.local_share_rel_pos_bias() if i not in \
```
Comment (ZwwWayne) [inline mmdet/models/backbones/vit.py:484]: it will be easier to read if we simply use the 
```
if xxx
   xxx
else:
   xxx
```

File: mmdet/models/backbones/vit.py [hunk 9]
```diff
@@ -0,0 +1,529 @@
...
+    def forward(self):
+        relative_position_bias = self.relative_position_bias_table[
+            self.relative_position_index.view(-1)].view(
+                self.window_size[0] * self.window_size[1] + 1,
+                self.window_size[0] * self.window_size[1] + 1, -1)
+        return relative_position_bias.permute(2, 0, 1).contiguous()
+
+
+@BACKBONES.register_module()
+class VisionTransformer(_VisionTransformer):
```
Comment (ZwwWayne) [inline mmdet/models/backbones/vit.py:323]: unit tests.

File: configs/mask_rcnn/mask_rcnn_vit_base_fpn_coco.py (hunk 5 above)
Comment (hhaAndroid) [reply configs/mask_rcnn/mask_rcnn_vit_base_fpn_coco.py:1]: https://github.com/open-mmlab/mmdetection/blob/master/configs/strong_baselines/README.md

File: mmdet/models/backbones/vit.py [hunk 10]
```diff
@@ -0,0 +1,529 @@
...
+        W_ = math.ceil(W / self.window_size) * self.window_size
+
+        qkv = self.qkv(x).transpose(1, 2).reshape(B, 3 * C, H, W)
+        qkv = F.pad(qkv, [0, W_ - W, 0, H_ - H], mode=self.pad_mode)
+
+        qkv = F.unfold(
+            qkv,
+            kernel_size=(self.window_size, self.window_size),
+            stride=(self.window_size, self.window_size))
+        B, C_kw_kw, L = qkv.shape
```
Comment (jshilong) [inline mmdet/models/backbones/vit.py:217]: C_kw_kw means?
